### PR TITLE
[PW-3771] Revise Recurring settings

### PR DESCRIPTION
**Pull request by AlexandrosMor**
**Description**

The recurring settings are modified for api version 66 based on the Adyen's [documentation](https://docs.adyen.com/online-payments/tokenization/create-and-use-tokens). In the admin page in magento there are two settings Vault enabled and Enable oneclick. When the Vault enabled is set then it is similar to the previous enable recurring and the following parameters are passed:
recurringProcessingModel : Subscription 
storedPaymentMethod: true 
shopperInteraction: ecommerce
If the Vault enabled and the enable oneclick is set then the one click is possible:
recurringProcessingModel : CardOnFile
enableOneclick: true 
shopperInteraction: ecommerce
but the storedPaymentMethod is not included in the request.

Instructions how to set the Recurring setting in customer are in Adyen's docs [here](https://docs.adyen.com/online-payments/tokenization/create-and-use-tokens#test-and-go-live) This setting will always enable the card to be stored in Adyen's backend

In Magento2 there are new two fields in admin panel under the stored payment methods
Vault 
OneClick

The vault enables the stored token in magento and the recurringProcessingModel is Subscription. The oneclick enables the stored pyament method in adyen backend for later transactions with the shopper presence and the recurringProcessingModel is CardOnFile. When both are enabled Vault is override the oneclick and the recurringProcessingModel is Subscription again.

**Tested scenarios**

Screenshot 2021-03-09 at 08 57 30

**Fixed issue**:

### CODE REVIEW

File: Gateway/Response/VaultDetailsHandler.php [hunk 1]
```diff
@@ -55,9 +64,16 @@ public function handle(array $handlingSubject, array $response)
         /** @var PaymentDataObject $orderPayment */
         $orderPayment = SubjectReader::readPayment($handlingSubject);
         $additionalData = $orderPayment->getPayment()->getAdditionalInformation();
+
+        if ($this->adyenHelper->isCreditCardVaultEnabled()) {
+            $this->vaultHelper->saveRecurringDetails($orderPayment->getPayment(), $response['additionalData']);
+        }
+
+        /*
+         * Token enabler is only sent from the admin payment while we should not rely on such a checkbox
```
Comment (cyattilakiss) [inline Gateway/Response/VaultDetailsHandler.php:73]: @AlexandrosMor could you pls create a ticket to clean up the is_active_payment_token_enabler and start using the storePaymentMethod also in the admin if still applicable?

**Review by cyattilakiss — COMMENTED** [review]
Is the docs suggestion created for the changed configuration page?

**Review by cyattilakiss — APPROVED** [review]

**Review by peterojo — APPROVED** [review]

File: Gateway/Response/VaultDetailsHandler.php (hunk 1 above)
Comment (AlexandrosMor) [reply Gateway/Response/VaultDetailsHandler.php:73]: The is_active_payment_token_enabler is checked by the magento in order to save the token. Magento is checking 
```
 const IS_ACTIVE_CODE = 'is_active_payment_token_enabler';
and is used in magento: 

$paymentToken->setIsVisible(
            (bool) (int) ($additionalInformation[VaultConfigProvider::IS_ACTIVE_CODE] ?? 0)
        );
        ```

**Comment by cyattilakiss** [comment]
Please check sonarcloud code smells!

**Review by cyattilakiss — APPROVED** [review]

File: Gateway/Request/OneclickAuthorizationDataBuilder.php [hunk 2]
```diff
@@ -69,13 +69,9 @@ public function build(array $buildSubject)
...
 
-        if ($payment->getAdditionalInformation('customer_interaction')) {
-            $shopperInteraction = "Ecommerce";
-        } else {
-            $shopperInteraction = "ContAuth";
-        }
-
-        $requestBody['shopperInteraction'] = $shopperInteraction;
+        // We override the previously set shopperInteraction in the buildRecurringData() to ContAuth here because if
+        // it's a later payment we should always send ContAuth
```
Comment (acampos1916) [inline Gateway/Request/OneclickAuthorizationDataBuilder.php:73]: Just for more clarity 
```suggestion
        // it's a payment with a stored method we should always send ContAuth
```

File: Gateway/Response/VaultDetailsHandler.php [hunk 3]
```diff
@@ -54,10 +63,16 @@ public function handle(array $handlingSubject, array $response)
...
         $tokenEnabler = array_key_exists("is_active_payment_token_enabler", $additionalData) ? $additionalData['is_active_payment_token_enabler']:false;
         if ($tokenEnabler) {
             $this->vaultHelper->saveRecurringDetails($orderPayment->getPayment(), $response['additionalData']);
-        }
+        }*/
```
Comment (acampos1916) [inline Gateway/Response/VaultDetailsHandler.php:76]: Should we remove this commented out code?

File: Helper/Requests.php [hunk 4]
```diff
@@ -374,40 +374,37 @@ public function buildRedirectData($storeId, $request = [])
...
-            $request['enableOneClick'] = $enableOneclick && !$isGuestUser && $shouldStoreCreditCardInfo;
-            $request['enableRecurring'] = (bool)$enableRecurring;
+        $enableOneclick = $this->adyenHelper->getAdyenAbstractConfigData('enable_oneclick', $storeId);
+        $enableVault = $this->adyenHelper->isCreditCardVaultEnabled();
 
-            // value can be 0,1 or true
-            if ($shouldStoreCreditCardInfo || ($isGuestUser && $this->adyenHelper->isGuestTokenizationEnabled($storeId))) {
-                $request['paymentMethod']['storeDetails'] = true;
+        // TODO Remove it in version 7
+        if ($payment->getAdditionalInformation(AdyenCcDataAssignObserver::STORE_CC)) {
```
Comment (acampos1916) [inline Helper/Requests.php:389]: Do we have a ticket so that we don't forget to change this? Or not needed?

Comment (AlexandrosMor) [reply Helper/Requests.php:389]: good idea lets create a ticket for it :)

**Comment by acampos1916** [comment]
@AlexandrosMor I think the test scenarios picture is missing a couple of conditions, maybe just a typo or not needed? But just to make sure:

 * Vault: no | OneClick: no | Settings in the Back office: false.
 * Vault: no | OneClick: no | Settings in the Back office: true.

Also, what's the name of the "Settings in the Back office"? Since merchants have access to Customer Area I think it is a better idea to show the name/path to the CA setting instead?

**Comment by AlexandrosMor** [comment]
> @AlexandrosMor I think the test scenarios picture is missing a couple of conditions, maybe just a typo or not needed? But just to make sure:
> 
> * Vault: no | OneClick: no | Settings in the Back office: false.
> * Vault: no | OneClick: no | Settings in the Back office: true.
> 
> Also, what's the name of the "Settings in the Back office"? Since merchants have access to Customer Area I think it is a better idea to show the name/path to the CA setting instead?

The description is updated

**Review by acampos1916 — DISMISSED** [review]

**Review by cyattilakiss — DISMISSED** [review]

File: Helper/Requests.php [hunk 5]
```diff
@@ -374,40 +374,37 @@ public function buildRedirectData($storeId, $request = [])
...
+            $request['storePaymentMethod'] = true;
+        }
+        //recurring
+        if ($enableVault ) {
+
```
Comment (msilvagarcia) [inline Helper/Requests.php:394]:
```suggestion
        if ($enableVault) {
```

File: Helper/Requests.php [hunk 6]
```diff
@@ -374,40 +374,37 @@ public function buildRedirectData($storeId, $request = [])
...
+            }else{
+                $request['recurringProcessingModel'] = 'Subscription';
             }
         }
-
         return $request;
     }
 
+
+
```
Comment (msilvagarcia) [inline Helper/Requests.php:407]:
```suggestion
```

**Review by msilvagarcia — APPROVED** [review]
There are some great improvements to the code here. I like what I see :)

**Review by msilvagarcia — APPROVED** [review]

**Pull request merged**